model/src/main/kotlin/io/spine/validation/ErrorMessage.kt

```diff
@@ -71,27 +84,68 @@ private constructor(private val value: String) {
          */
         @JvmStatic
         @JvmOverloads
-        public fun forComposite(format: String,
-                         left: ErrorMessage,
-                         right: ErrorMessage,
-                         operation: LogicalOperator = BO_UNKNOWN): ErrorMessage =
-            ErrorMessage(
-                format
-                    .replacePlaceholder(LEFT, left.value)
-                    .replacePlaceholder(RIGHT, right.value)
-                    .replacePlaceholder(OPERATION, operation.printableString())
-            )
+        public fun forComposite(
+            format: String,
+            left: ErrorMessage,
+            right: ErrorMessage,
+            operation: LogicalOperator = LO_UNKNOWN,
+            value: String = ""
+        ): ErrorMessage {
+            val msg = Template(format).apply {
+                formatStatic(OPERATION, operation.printableString())
+                formatDynamic(LEFT, left.expression)
+                formatDynamic(RIGHT, right.expression)
+                if (value.isNotBlank()) {
+                    formatDynamic(VALUE, value)
+                }
+            }
+            return ErrorMessage(msg.joinExpression())
+        }
     }
 
-    override fun toString(): String = value
-}
-
-private fun String.replacePlaceholder(placeholder: String, newValue: String): String {
-    val formattedPlaceholder = "{$placeholder}"
-    return replace(formattedPlaceholder, newValue)
+    override fun toString(): String = expression
 }
 
-private fun LogicalOperator.printableString() = when(this) {
+private fun LogicalOperator.printableString() = when (this) {
     AND, OR, XOR -> name.lowercase()
     else -> "<unknown operation>"
 }
+
+/**
+ * A placeholder which can be present in a validation error message.
+ */
+internal enum class Placeholder {
+
+    /**
+     * The actual value of the validated field.
+     */
+    VALUE,
+
+    /**
+     * The value to which the validated field is compared.
+     */
+    OTHER,
+
+    /**
+     * In a composite validation rule, the left condition.
+     */
+    LEFT,
+
+    /**
+     * In a composite validation rule, the right condition.
+     */
+    RIGHT,
+
+    /**
+     * In a composite validation rule, the boolean operator which joins the two conditions.
+     *
+     * @see LogicalOperator
+     */
+    OPERATION;
```

Comment:
The description and `@see` says "operator". I would name this one `OPERATOR` as well.